### ISSUE TITLE
Add support for Image Semantic Segmentation

### ISSUE THREAD

**Issue by jwmueller** [cleanlab/cleanlab#376]
Previous requests:

https://github.com/cleanlab/cleanlab/issues/38

https://github.com/cleanlab/cleanlab/issues/30

https://github.com/cleanlab/cleanlab/issues/26

**Comment by FrankwaP** [comment]
Hi!

How is the implementation going?
I'm trying to "feed" semantic segmentation labels and predictions to cleanlab. I'm using `reshape(-1, class_number)` in my image. And will use `reshape(N, class_number, H, W)` on the output, to bring it back in "image referential". Then I'll try raising the image with the highest percentage of incorrect pixels, or maybe work at class level… we'll see…

Maybe I can make myself useful to you in the meantime ?

**Comment by jwmueller** [comment]
Awesome to hear! We'll have some thoughts on this for you shortly. 
In the meantime, if you'd like more interactive discussion, feel free to join our slack:
https://cleanlab.ai/slack

but chatting here in github is fine as well if you prefer!

**Comment by FrankwaP** [comment]
Fine, I'll wait for your input then. Since you don't mind, I'd rather keep the discussion on github thank you :-)

**Comment by vdlad** [comment]
Hi @FrankwaP , I could use a hand with semantic segmentation for cleanlab. To test the error detection accuracy of an algorithm, a reliable dataset is crucial. I've reviewed the test set, but it would be great to have a second pair of eyes.

Please check the Google Sheets at this [link](https://docs.google.com/spreadsheets/d/1Id9GyIVmgOsBl28x5FBqAiQDu9AMqO3Z4a4fcjJKADg/edit?usp=sharing)

There are three pages:

1. The first page allows you to mark if an image contains an error and add an error location.

2. The second page has official cityscapes annotation rules with detailed labeling policies when you hover over the class.

3. The third page shows an html of a training notebook with four images. Column 1: RGB image, Column 2: official mask overlayed on the image, Column 3: official mask, Column 4: pretrained model classification.
Note: Ignore the car hood in all images. Thanks for your help!

**Comment by jwmueller** [comment]
To clarify, the goal of ^ is to help us strengthen our benchmark of label error detection algorithms. Our goal is to manually assign a `has_error` boolean value to each image based on whether or not its annotated segmentation mask contains any sort or error or not.  We have done this for the Cityscapes test set, but a second pair of eyes would be very helpful. Based on this ground truth `has_error` information, we are evaluating various methods for label error detection in semantic segmentation to decide which one is best to add to clearnlab, so it's crucial for us that the ground truth be accurate.

A second task you could be helpful with after that is: to train 2 different models on your favorite public segmentation dataset and save their out-of-sample predicted probabilities (ideally a popular dataset in ML/vision research). Then we can run our label error detection  algorithms on your dataset as well (we want at least 2 different models to make sure our methods are effective with different types of models) to see what sorts of label errors different algorithms do/don't reveal.

**Comment by FrankwaP** [comment]
I sure can help with the Cityscape dataset, but I'm wondering if a synthetic dataset could be more useful to get a robust testing of cleanlab algorithms, as we would go from 100% correct annotations then it would allow to precisely control what and where the label errors are (class mislabeling or incorrect annotation shapes). 

I'm think of a dataset such as [SYNTHIA](https://paperswithcode.com/paper/the-synthia-dataset-a-large-collection-of) which is similar to Cityscape. But it could be a very simple and stupid one, generated using few commands from PIL/scikit-image to generate various shapes.

What's your take on it? Have you already discussed this option?

**Comment by vdlad** [comment]
Thanks @FrankwaP  ! Our SYNTHIA dataset is up and running, with various synthetic mask errors included. Our aim with Cityscapes is to simulate a real-world scenario for a data scientist using Cleanlab to detect errors in their own dataset. By having realistic natural errors and trained models, we can evaluate the performance of error detection algorithms.

**Comment by jwmueller** [comment]
@FrankwaP as @vdlad mentioned we are already benchmarking using SYNTHIA, but it's a great suggestion!
After you help us improve our CityScapes dataset, a third dataset would be great to have.

Option 1:  As I mentioned one option for this would be to just train 2 different models on your favorite public segmentation dataset and save their out-of-sample predicted probabilities (ideally a popular dataset in ML/vision research). Then we can run our label error detection algorithms on your dataset as well (we want at least 2 different models to make sure our methods are effective with different types of models) to see what sorts of label errors different algorithms do/don't reveal. We won't have ground truth `has_error` info for your dataset, but we can nonetheless look over say the top-100 images flagged as most likely to have label error and manually count how many of them actually do (precision@100 evaluation).

Option 2:  Another synthetic segmentation dataset like SYNTHIA where you can manually introduce label errors if know of another good one? This would allow for comprehensive evaluation of algorithms like we are doing with SYNTHIA & CityScapes, but the downside ofc is the conclusions depend on how we decided to introduce the label errors (which is why CityScapes evaluation is valuable, because it is based on naturally-occurring label errors whose distribution may look different than ones we synthetically introduce).
